dangerzone/conversion/pixels_to_pdf.py

```diff
@@ -27,137 +29,72 @@ async def convert(
-            pdf_filename = f"{tempdir}/page-{page}.pdf"
 
             with open(width_filename) as f:
-                width = f.read().strip()
+                width = int(f.read().strip())
             with open(height_filename) as f:
-                height = f.read().strip()
-
+                height = int(f.read().strip())
+            with open(rgb_filename, "rb") as rgb_f:
+                untrusted_rgb_data = rgb_f.read()
             # The first few operations happen on a per-page basis.
             page_size = os.path.getsize(filename_base + ".rgb") / 1024**2
             total_size += page_size
             timeout = self.calculate_timeout(page_size, 1)
 
+            self.update_progress(
+                f"Converting page {page_num}/{num_pages} from pixels to PDF"
+            )
+            pixmap = fitz.Pixmap(
+                fitz.Colorspace(fitz.CS_RGB), width, height, untrusted_rgb_data, False
+            )
             if ocr_lang:  # OCR the document
-                self.update_progress(
-                    f"Converting page {page}/{num_pages} from pixels to searchable PDF"
-                )
-                await self.run_command(
-                    [
-                        "gm",
-                        "convert",
-                        "-size",
-                        f"{width}x{height}",
-                        "-depth",
-                        "8",
-                        f"rgb:{rgb_filename}",
-                        f"png:{png_filename}",
-                    ],
-                    error_message=f"Page {page}/{num_pages} conversion to PNG failed",
-                    timeout_message=(
-                        "Error converting pixels to PNG, convert timed out after"
-                        f" {timeout} seconds"
-                    ),
-                    timeout=timeout,
-                )
-                await self.run_command(
-                    [
-                        "tesseract",
-                        png_filename,
-                        ocr_filename,
-                        "-l",
-                        ocr_lang,
-                        "--dpi",
-                        "70",
-                        "pdf",
-                    ],
-                    error_message=f"Page {page}/{num_pages} OCR failed",
-                    timeout_message=(
-                        "Error converting PNG to searchable PDF, tesseract timed out"
-                        f" after {timeout} seconds"
-                    ),
-                    timeout=timeout,
-                )
-
+                try:
+                    ocr_pdf_bytes = pixmap.pdfocr_tobytes(
+                        compress=True,
+                        language=ocr_lang,
+                        tessdata=get_tessdata_dir(),
+                    )
+                except:  # containers' verison has old API without tessdata
+                    ocr_pdf_bytes = pixmap.pdfocr_tobytes(
+                        compress=True,
```

Comment:
How does this `compress=True` argument work with `deflate_images=True` later on? Is it redundant, can we forego one or the other?

Reply:
That's amazing! The minimal time overhead (in this case, 100ms per page) and the size savings in contrast to `ps2pdf` are very promising. Regarding my original question, should we disable `deflate_images` once compression is enabled for OCR?

Reply:
You know what, all I think is that if `deflate_images=True` does not pose an extra time overhead, and the bytes are exactly the same, this would mean that the `mupdf` code auto-detects that the content stream is in DEFLATE format, and does not redo the compression. Looking into the code though, this does not seem to be the case:

https://github.com/ArtifexSoftware/mupdf/blob/master/source/pdf/pdf-write.c#L2133C2-L2144C2

On the other hand, I tried to reproduce this by doing a test on my own. I OCRed a page with `compress=True`, and then saved the pdf with `deflate_images` either set to True or False. The result does have the exact same number of bytes, so your measurements are indeed correct!

_(The SHA-256 hash is different, but that's only due to a different ID value in the PDF trailer. Other than that, the content streams are byte-to-byte equal)_

---

Worst-case scenario, turns out we actually do double conversion, but its overhead is minor compared to the overhead of the OCR. I can't prove this though, so here's what we can do here. Let's add a note and say that we enable both flags, because we have measured their impact and we have not seen a noticeable overhead. Let's even link to this discussion here, for future reference. This will simplify the code and we will be able to refer to this decision later on. Are you ok with that?